source/pele/cell_lists.h

```diff
@@ -150,40 +200,273 @@ namespace pele {
  */
-template <class visitor_t>
+template <class visitor_t, size_t ndim>
 class CellListsLoop {
 protected:
     visitor_t & m_visitor;
-    pele::Array<long> const m_ll;
-    pele::Array<long> const m_hoc;
-    std::vector<std::pair<size_t, size_t> > const & m_cell_neighbor_pairs;
+    CellListsContainer<ndim> const & m_container;
 
 public:
-    CellListsLoop(visitor_t & visitor, CellListsContainer const & container)
+    CellListsLoop(visitor_t & visitor, CellListsContainer<ndim> const & container)
         : m_visitor(visitor),
-          m_ll(container.m_ll),
-          m_hoc(container.m_hoc),
-          m_cell_neighbor_pairs(container.m_cell_neighbor_pairs)
+          m_container(container)
     {}
 
-
     void loop_through_atom_pairs()
     {
-        for (auto const & ijpair : m_cell_neighbor_pairs) {
+        typename CellListsContainer<ndim>::const_iterator iiter, jiter, iend, jend;
+        iend = m_container.end();
+        for (auto const & ijpair : m_container.m_cell_neighbor_pairs) {
             const size_t icell = ijpair.first;
             const size_t jcell = ijpair.second;
             // do double loop through atoms, avoiding duplicate pairs
-            for (auto iiter = AtomInCellIterator(m_ll.data(), m_hoc[icell]); !iiter.done(); ++iiter) {
-                size_t const atomi = *iiter;
+            for (iiter = m_container.begin(icell); iiter != iend; ++iiter) {
                 // if icell==jcell we need to avoid duplicate atom pairs
-                long const loop_end = (icell == jcell) ? atomi : CELL_END;
-                for (auto jiter = AtomInCellIterator(m_ll.data(), m_hoc[jcell], loop_end); !jiter.done(); ++jiter) {
-                    size_t const atomj = *jiter;
-                    m_visitor.insert_atom_pair(atomi, atomj);
+                jend = (icell == jcell) ? iiter : m_container.end();
+                for (jiter = m_container.begin(jcell); jiter != jend; ++jiter) {
+                    m_visitor.insert_atom_pair(*iiter, *jiter);
                 }
             }
         }
+    }
+};
+
+
+template<typename distance_policy>
+class LatticeNeighbors {
+public:
+    static const size_t ndim = distance_policy::_ndim;
+    std::shared_ptr<distance_policy> m_dist; // the distance function
+
+    typedef VecN<ndim> cell_vec_t;
+    pele::Array<double> m_boxvec;
+    double m_rcut;
+    cell_vec_t m_ncells_vec; // the number of cells in each dimension
+    pele::VecN<ndim> m_rcell_vec; // the cell length in each dimension
+    size_t m_ncells;
+
+    LatticeNeighbors(std::shared_ptr<distance_policy> dist,
+            pele::Array<double> boxvec,
+            double rcut,
+            pele::Array<size_t> ncells_vec)
+        : m_dist(dist),
+          m_boxvec(boxvec.copy()),
+          m_rcut(rcut),
+          m_ncells_vec(ncells_vec.begin(), ncells_vec.end())
+    {
+        m_ncells = 1;
+        for (size_t idim = 0; idim < ndim; ++idim) {
```

Comment:
Could this be done with the divide operator of the vector class?
And equally with the prod() function?